ADR: Solution Architecture (#169)
nuts-foundation/nuts-knooppunt: docs/adr/003-knooppunt-solution-architecture.md

- **Flexibility**: Can be adapted to different deployment environments and use cases.
- **Modularity**: Components (e.g. Nuts node, Open Policy Agent) can be enabled/disabled as needed.
- **Security**: Minimize attack surfaces.
- **Pluggability**: Should be as easy as possible to integrate.

### Use cases

Deciding on the architecture depends on the use cases we want to support. The main use cases are:

- Handling data exchange, initiated by a remote party (inbound)
  - This involves receiving requests, authenticating, authorizing, and responding.
  - It could also involve masking or filtering data based on consent or authorization rules, before it's returned to the requester.
- Initiating data exchange with a system from a remote party (outbound)
    - This involves looking up parties, endpoints, authentication
    - Outbound requests are out of scope for this ADR.

## Considered Options
This section describes considered architecture options.

In any of the options, there's a proxy (e.g. NGINX, HAProxy, Traefik) in front of the Knooppunt and/or EHR FHIR API.
This is a typical reverse proxy, handling TLS termination, routing, load balancing, caching, etc.

> ### Knooppunt as both PEP and PDP

Reviewer: ```suggestion
### Option 1: Knooppunt as both PEP and PDP
```

The Knooppunt sits on the edge of the vendor's network, handling all data exchanges. It acts as:
- OAuth2 Authorization Server
- Policy Decision Point
- Policy Enforcement Point

```text
┌─────────────────┐       ┌─────────┐       ┌────────────┐       ┌──────────────┐
│                 │       │         │       │            │       │              │
│ External System ├──────►│  Proxy  │──────►│ Knooppunt  ├──────►│ EHR FHIR API │
│                 │       │         │       │ [PEP/PDP]  │       │              │
└─────────────────┘       └─────────┘       └────────────┘       └──────────────┘
```

> Data exchanges are routed through the Knooppunt, theoretically offloading all "complicated" concerns from the vendor.

Reviewer: ```suggestion
All inbound data requests are routed through the Knooppunt, theoretically offloading all "complicated" concerns from the vendor.
```

- Advantages:
  - Simpler deployment, since it only requires the Knooppunt to be deployed.
  - Easier to configure, since all configuration is centralized in the Knooppunt.
  - Easier to reason about, since all data exchanges go through a single component.
- Disadvantages:
  - Large attack surface on the Knooppunt, since it handles all data exchanges.
    Makes it harder to use security measures vendors already have in place, especially for resource transformation and filtering.

>   - Might not actually make things easier for vendors, if they want to implement requirements not supported by the Knooppunt (e.g. auditing, data minimization)

Reviewer: ```suggestion
- Advantages:
  - Simple deployment, since it only requires the Knooppunt to be deployed.
  - Easy to configure, since all configuration is centralized in the Knooppunt.
  - Easy to reason about, since all data exchanges go through a single component.
- Disadvantages:
  - Large attack surface on the Knooppunt, since it handles all inbound requests.
    Makes it harder to use security measures vendors already have in place, especially for resource transformation and filtering.
  - Less flexible when the vendor needs additional concerns not supported by the Knooppunt (e.g. auditing, data minimization)
```

> ### Knooppunt as PDP, separate PDP

Reviewer: ```suggestion
### Option 2: Knooppunt as PDP, separate PDP
```

  - Easier to integrate with existing security infrastructure (e.g. existing reverse proxies).
  - More flexibility for vendors to choose or reuse a Policy Enforcement Point that fits their needs.
  - Easier to align with vendor compliancy requirements.
- Disadvantages:
  - More complex deployment, since it requires an additional component (the proxy).
  - More complex configuration, since the proxy needs to be set up correctly to work with the Knooppunt.
  - Potentially more points of failure, since there are more components involved.

### Variant for outbound data exchanges
Outbound data exchanges, initiated by the local EHR, could be routed through the Knooppunt. This could offload the EHR from:

- Negotiating TLS with the external party
- Negotiating authentication with the external party
- Looking up the right endpoint to call

This is not in scope for this ADR.

## Decision Outcome

We have decided to implement the Knooppunt as a Policy Decision Point (PDP), option "Knooppunt as internal system";
integrating with existing infrastructure via an external proxy for inbound data exchanges.

> This approach was chosen because it minimizes the attack surface of the Knooppunt, leverages existing security components, and provides flexibility for vendors to use their preferred Policy Enforcement Point (PEP).

Author: According to the reference implementation, that follows XACML, the PDP should retrieve whatever additional info it needs for decisions: https://en.wikipedia.org/wiki/XACML#/media/File:XACML_Architecture_&_Flow.png

> This approach was chosen because it minimizes the attack surface of the Knooppunt, leverages existing security components, and provides flexibility for vendors to use their preferred Policy Enforcement Point (PEP).

Reviewer: Ok, that makes sense. Then I agree with the simplicity of the PEP.
